src/oidc_child/oidc_child_curl.c

```diff
@@ -0,0 +1,462 @@
+                /* Older Azure AD v1 endpoints expect 'code' instead of the RFC
+                 * conforming 'device_code', see e.g.
+                 * https://docs.microsoft.com/de-de/archive/blogs/azuredev/assisted-login-using-the-oauth-deviceprofile-flow
+                 * and search for 'request_content' in the code example. */
+                talloc_free(post_data);
+                post_data = talloc_asprintf(mem_ctx, post_data_tmpl, client_id, "code",
+                                                                     dc_ctx->device_code);
+                if (post_data == NULL) {
+                    DEBUG(SSSDBG_OP_FAILURE, "Failed to generate POST data.\n");
+                    ret = ENOMEM;
+                    goto done;
+                }
+                azure_fallback = true;
+                continue;
+            }
+            break;
+        }
+
+        /* only run once after getting the device code to tell the IdP we are
+         * expecting that the user will connect */
+        if (get_device_code) {
+            if (ret == EAGAIN) {
+                ret = EOK;
+            }
+            break;
+        }
+
+        sleep(dc_ctx->interval);
+        waiting_time += dc_ctx->interval;
+    } while (waiting_time < dc_ctx->expires_in);
+
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to get token.\n");
+    }
+
+done:
+    talloc_free(post_data);
+    talloc_free(error_description);
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl_ctx);
+    return ret;
+}
+
+errno_t get_openid_configuration(struct devicecode_ctx *dc_ctx,
+                                        const char *issuer_url)
+{
+    int ret;
+    char *uri = NULL;
+    bool has_slash = false;
+
+    if (issuer_url[strlen(issuer_url) - 1] == '/') {
+        has_slash = true;
+    }
+
+    uri = talloc_asprintf(dc_ctx, "%s%s.well-known/openid-configuration",
+                                   issuer_url, has_slash ? "" : "/");
+    if (uri == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate memory for results.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    clean_http_data(dc_ctx);
+    ret = do_http_request(dc_ctx, uri, NULL, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "http request failed.\n");
+    }
+
+done:
+    talloc_free(uri);
+
+    return ret;
+}
+
+#define DEFAULT_SCOPE "user"
+
+errno_t get_devicecode(struct devicecode_ctx *dc_ctx,
+                       const char *client_id)
+{
+    int ret;
+
+    char *post_data = NULL;
+
+    post_data  = talloc_asprintf(dc_ctx, "client_id=%s&scope=%s",
+                                 client_id,
+                                 dc_ctx->scope != NULL ? dc_ctx->scope
+                                                       : DEFAULT_SCOPE);
+    if (post_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate memory for POST data.\n");
+        return ENOMEM;
+    }
+
+    clean_http_data(dc_ctx);
```

Comment:
It looks like repeated before every(?) `do_http_request()`?
Would it make sense to move there?

Reply:
I would like to keep it this way so that `do_http_request()` can append data to an existing buffer, if needed.